packages/core/android/src/main/kotlin/com/datadog/reactnative/DdSdkBridgeExt.kt

```diff
@@ -6,44 +6,187 @@
+                it.toWritableArray(
+                    createWritableMap,
+                    createWritableArray
+                )
+            )
+            is Map<*, *> -> writableArray.pushMap(
+                it.toWritableMap(
+                    createWritableMap,
+                    createWritableArray
+                )
+            )
+            else -> Log.e(
+                javaClass.simpleName,
+                "Unhandled type ${it.javaClass.simpleName} has been ignored"
+            )
         }
     }
-    return list
+
+    return writableArray
 }
 
-internal fun Map<*, *>.toWritableMap(): WritableNativeMap {
-    val map = WritableNativeMap()
-    for ((k,v) in iterator()) {
+/**
+ * Converts the [Map] to a [WritableNativeMap].
+ */
+internal fun Map<*, *>.toWritableMap(): WritableMap {
+    return this.toWritableMap(
+        createWritableMap = { WritableNativeMap() },
+        createWritableArray = { WritableNativeArray() }
+    )
+}
+
+/**
+ * Converts the [Map] to a [WritableMap].
+ * @param createWritableMap a function to provide a concrete instance for WritableMap(s)
+ * @param createWritableArray a function to provide a concrete instance for WritableArray(s)
+ */
+internal fun Map<*, *>.toWritableMap(
+    createWritableMap: () -> WritableMap,
+    createWritableArray: () -> WritableArray
+): WritableMap {
+    val map = createWritableMap()
+
+    for ((k, v) in iterator()) {
         val key = (k as? String) ?: k.toString()
-        @Suppress("NotImplementedDeclaration")
         when (v) {
             null -> map.putNull(key)
             is Int -> map.putInt(key, v)
             is Long -> map.putDouble(key, v.toDouble())
             is Float -> map.putDouble(key, v.toDouble())
             is Double -> map.putDouble(key, v)
             is String -> map.putString(key, v)
-            is List<*> -> map.putArray(key, v.toWritableArray())
-            is Map<*, *> -> map.putMap(key, v.toWritableMap())
-            else -> TODO()
+            is Boolean -> map.putBoolean(key, v)
+            is List<*> -> map.putArray(
+                key,
+                v.toWritableArray(
+                    createWritableMap,
+                    createWritableArray
+                )
+            )
+            is Map<*, *> -> map.putMap(
+                key,
+                v.toWritableMap(
+                    createWritableMap,
+                    createWritableArray
+                )
+            )
+            else -> Log.e(
+                javaClass.simpleName,
+                "Unhandled type ${v.javaClass.simpleName} has been ignored"
+            )
         }
     }
+
     return map
 }
+
+/**
+ * Recursively converts the [ReadableMap] to a [Map] which only contains Kotlin stdlib objects,
+ * such as [List], [Map] and the raw types.
+ */
+internal fun ReadableMap.toMap(): Map<String, Any> {
+    val map: HashMap<String, Any> = this.toHashMap()
+    val iterator: Iterator<String> = map.keys.iterator()
```

Comment:
minor: explicit types can be removed unless compiler infers them in a wrong way; they are not very useful

```suggestion
    val map = this.toHashMap()
    val iterator = map.keys.iterator()
```